addons/website_sale_delivery/models/sale_order.py

```diff
@@ -1,192 +1,146 @@
+            self._delivery_unset()
             return True
-        else: 
-            carrier_id = force_carrier_id or order.carrier_id.id
-            carrier_ids = self._get_delivery_methods(cr, uid, order, context=context)
+        else:
+            carrier_id = force_carrier_id or self.carrier_id.id
+            carrier_ids = self._get_delivery_methods()
             if carrier_id:
                 if carrier_id not in carrier_ids:
                     carrier_id = False
                 else:
                     carrier_ids.remove(carrier_id)
                     carrier_ids.insert(0, carrier_id)
-            if force_carrier_id or not carrier_id or not carrier_id in carrier_ids:
-                for delivery_id in carrier_ids:
-                    carrier = carrier_obj.verify_carrier(cr, SUPERUSER_ID, [delivery_id], order.partner_shipping_id)
+            if force_carrier_id or not carrier_id or carrier_id not in carrier_ids:
+                for delivery in self.env['delivery.carrier'].browse(carrier_ids):
+                    carrier = delivery.verify_carrier(self.partner_shipping_id)
                     if carrier:
-                        carrier_id = delivery_id
+                        carrier_id = delivery.id
                         break
-                order.write({'carrier_id': carrier_id})
+                self.write({'carrier_id': carrier_id})
             if carrier_id:
-                order.delivery_set()
+                self.delivery_set()
             else:
-                order._delivery_unset()
+                self._delivery_unset()
 
         return bool(carrier_id)
 
-    def _get_delivery_methods(self, cr, uid, order, context=None):
-        carrier_obj = self.pool.get('delivery.carrier')
-        carrier_ids = carrier_obj.search(cr, SUPERUSER_ID, [('website_published', '=', True)], context=context)
+    def _get_delivery_methods(self):
+        self.ensure_one()
+        carriers = self.env['delivery.carrier'].with_context(order_id=self.id).search([('website_published', '=', True)])
         available_carrier_ids = []
         # Following loop is done to avoid displaying delivery methods who are not available for this order
         # This can surely be done in a more efficient way, but at the moment, it mimics the way it's
         # done in delivery_set method of sale.py, from delivery module
-
-        new_context = dict(context, order_id=order.id)
-        for carrier in carrier_ids:
-
+        for carrier in carriers:
             try:
-                _logger.debug("Checking availability of carrier #%s" % carrier)
-                available = carrier_obj.read(cr, SUPERUSER_ID, [carrier], fields=['available'], context=new_context)[0]['available']
-                if available:
-                    available_carrier_ids = available_carrier_ids + [carrier]
+                _logger.debug("Checking availability of carrier #%s" % carrier.id)
+                if carrier.available:
```

Comment:
@rga-odoo the point of this loop with try/except is *not* to make a global browse but read the records one by one